metagenomics.py

```diff
@@ -677,6 +677,40 @@ def taxa_hits_from_tsv(f, taxid_column=2):
     return c
 
 
+def kraken_leftover_diamond(krakenDb, diamondDb, taxDb, inBam, krakenReads, outReport, outReads, filterThreshold=None, threads=None):
+    with util.file.fifo(1) as sam_pipe:
+        proc = kraken_leftover_reads(krakenDb, inBam, krakenReads, sam_pipe, filter_threshold=filterThreshold)
+        diamond(sam_pipe, diamondDb, taxDb, outReport, outReads=outReads, threads=threads)
+        proc.wait()
+
+
+def kraken_leftover_reads(kraken_db, in_bam_fn, kraken_reads_fn, output_sam_fn, filter_threshold=None):
+    filter_threshold = filter_threshold if filter_threshold is not None else 0.05
+
+    env = os.environ.copy()
+    env['LC_ALL'] = 'C'
+
+    return subprocess.Popen('''
+    samtools view -h {bam} > {output} && \
```

Comment:
Okay ... a few comments here.
- this level of shell piping is making me uncomfortable (more than my previous comment below in the `diamond` method).
- at the very minimum it needs a `set -o pipefail`.. I've seen previous bits of code like this fail silently before I changed them
- are you sure it truly needs a sort before going into diamond? diamond doesn't require sorted input. If all you're trying to do is filter on unclassified input, that shouldn't require a sort, just a filter on taxid:32644.
- `kraken_leftover_reads` seems to be doing two things at once.. and it creates a weird user interface where the user is being asked for a kraken_db to run diamond on the leftover kraken reads, which isn't clear why that's necessary. It looks like the kraken_db is needed for `kraken-filter` (I'm not sure why, but oh well). Can we break up `kraken_leftover_reads` into `kraken_filter` and `kraken_unclassified` (or something like that) where the first step calls `kraken-filter` and the second does not require a kraken_db, and then `kraken_leftover_diamond` can pipe-glue them all together (if filterThreshold is not None... but if filterThreshold is None, just pipe-glue the first and last step and then you don't even need a kraken_db at all). In the most typical scenarios, I think users won't be messing with re-filtering much unless they're experimenting with things (since the normal kraken command includes a baseline filter) so any time we can avoid having to provide a 100GB db that would be great.
- separately, since we have our own fork of kraken, do you know if it's really necessary for `kraken-filter` to have a kraken_db as input?

Reply:
Broke it up to `kraken_leftover_reads` and doing both the leftover reads subset and diamond is now only in the snakemake rule.

It only needs the taxonomy files from the `kraken_db` so it doesn't load any of the big files. The goal is to use the kraken db's included taxonomy files to prevent any potential conflicts